Implement Vector

lthibault opened the pull request:
@spy16 I took a stab at a naive implementation of `builtin.Vector`.  It has a persistent interface, implemented via copy-on-write semantics.

If you have a moment, would you mind updating the `fn` call signature to use `builtin.Vector`?

⏱️ Estimated review time:  10 minutes
✅ Merge when ready

lthibault committed:
Implement Vector.

lthibault committed:
Simplify builtin.Vector.Conj

spy16 commented:
Nice. Will take a look today evening sometime if I get time.

spy16 commented on builtin/vector.go:
```diff
@@ -0,0 +1,33 @@
+
+var (
+	_ core.Any = (*Vector)(nil)
+	_ core.Seq = (*Vector)(nil)
```
Logically `Vector` is a sequence. But in our Sabre discussions, we had decided to make it a `Seqable` (i.e., a type with `Seq() (Seq, error)` method) and define a `Vector` interface to allow for multiple Vector implementations. Any reasons why you went with Vector type which itself is a `Seq`? 

I am fine with this approach by the way (In Sabre model, I would've said no to this since special forms parser for example needed to differentiate a vector from a list to allow for `(let [...] ....)` kind of forms. But now that we have a swappable `Analyzer` and an intermediate `Expr` which doesn't care how the values got there, means we can use this approach. Might even be slightly more performance because no interface boxing around actual types)... 

But `Vector` needs to support Assoc and index lookup operations along with normal `Seq` operations.

lthibault replied:
>Any reasons why you went with Vector type which itself is a Seq?

Honestly?  I had forgotten about our previous discussion, and I was in a rush to test something out... 😆 
I'll get this fixed up and ping you when it's ready.

>we had decided to make it a Seqable (

Can you remind me why we made this choice?  I gather from looking at https://github.com/spy16/sabre/pull/25 that the argument is basically the following:  lists are _defined_ by the `Seq` interface, so having `Vector` satisfy `Seq` would introduce some confusion.  For my own understanding, do you have an example of what this confusion would look like in practice?

spy16 replied:
> Can you remind me why we made this choice?

Consider the `(fn [a b c] <expr>*)` special form for example. If we don't have dedicated `Vector` interface, the parser would need to use `Seq` type-assertion to make sure second argument is a vector. But if we do that, `(fn (a b c) <expr>*)` would be valid as well, but it's not the syntax we expect. 

Since we were trying to support multiple implementations for a Vector (e.g., mutable vs immutable-persistent etc), this would have been a real problem in Sabre model because there was no intermediate `Expr` values and also no custom `Analyzer`.

But in the Slurp model, this problem is not that relevant, because we can get away with concrete-type assertions (i.e., `*LinkedList` when we want to make sure arg is a list and `*Vector` when we want it to be a vector).. This doesn't really limit us in terms of alternate vector implementations because that can still be done and a custom analyzer can be used.

lthibault replied:
Ok that makes sense, thanks!

lthibault commented:
TODO

- [x] `Vector` interface
- [x] `Seq` -> `Seqable`
- [x] `Assoc` method

lthibault commented:
@spy16 Do you have any objections if I start using [github.com/stretchr/testify/assert](https://github.com/stretchr/testify/assert) to manage unit tests?

It's a very minimal library (i.e. **not** a testing framework) that saves a lot of keystrokes.  Usage is basically:

```go
assert.Equal(t, want, got, msg)
```

spy16 commented:
I was simply trying to keep zero-external dependencies promise. But adding `testify` package should be okay I think. (But I would highly recommend **against** using `testify/mock` and `testify/suite` packages though)

lthibault commented:
> I was simply trying to keep zero-external dependencies promise. But adding `testify` package should be okay I think.

Yeah, that's why I asked.  I'll add a note in the README to the effect of "zero dependencies (other than for unit tests)".

> (But I would highly recommend against using testify/mock and testify/suite packages though)

Yes, 100% agreed.  (I wan't even aware those existed until now 😄 )

lthibault committed:
Implement Clojure-style bit-partitioned vector trie.

lthibault committed:
Implement transientVector.

lthibault commented:
@spy16 I went all-in and implemented a Clojure-style bit-partitioned vector trie.  Everything is finished except:

1. `Vector.Pop()`
2. A couple of transient-based optimizations

One thing to note is that the `node` type is effectively a `[]interface{}`.  The reason for this apparent weirdness is that we can't use slices if we want `PersistentVector` to be hashable (i.e. usable as a key in a map).  In order to work around this, I've rolled my own slice-like type, `node`, which uses concrete arrays.

I wanted to get some early feedback on this -- what do you think?

spy16 commented on builtin/vector_test.go:
```diff
@@ -0,0 +1,118 @@
+	}()
+
+	m := make(map[core.Vector]struct{})
+	m[EmptyVector] = struct{}{}
```
If we go with the `Hash() ([]byte, error)` model that we considered in Sabre, we would not need to make the type itself hash-able by Go runtime right? 🤔 

vector can provide a `Hash() ([]byte, error)` method that hash-based data structures make use of to generate the hash.. (internally vector can use any approach to do that -- but i am not sure what approaches are possible)..

lthibault replied:
> If we go with the Hash() ([]byte, error) model that we considered in Sabre, we would not need to make the type itself hash-able by Go runtime right? 🤔

Correct.  I actually think our mental models for `Hashable` diverged a while back, though. Let me explain. 😄 

I had initially proposed the following interface for `Hashable`:

```go
type interface Hashable {
    Hash() interface{}  // should probably be called Hashable(), actually...
}
```

The intent was for `Hash()` to provide a hashable representation of the value, i.e. one that can be consumed by the Go runtime's hash function (`builtin.nilinterhash`).  While `[]byte` is certainly a "hashable representation", there's a subtle distinction:  I wanted `Hash()` to be implemented _only_ for types that were not natively hashable in Go.  In other words, I think the data-structures in `builtin` should use the Go runtime's built-in hash function!

Here's an illustration of how the pieces fit together.  I think code is clearer than text, here!  https://play.golang.org/p/0-QsKSmQNWg

I think the confusion was introduced when I said that `Hashable` decouples the hashability of a type from the actual hash algorithm.  While the `Hashable` interface does indeed leave the choice of hash algorithm to the caller, I did **not** mean to imply we should maintain direct compatibility with [`hash.Hash`](https://golang.org/pkg/hash/#Hash)!  The conversion from `interface{}` to `[]byte` should still be left as an exercise to the user, either through serialization or via [something more "clever"](https://play.golang.org/p/TVJyJQVzgRy).  Note that this is how Go's `hash` library is designed as well -- it is the user's responsibility to marshal their data-structures into `[]byte` prior to hashing.

I therefore think it's fine -- and in fact even *preferable* -- for Slurp's internals to be tightly coupled to the Go runtime.  That is to say, the types in the `builtin` package should use Go's `nilinterhash` because it's going to be the most optimized, standardized hash function at our disposal.  My point about decoupling was simply that the `Hashable` interface avoids  imposing `nilinterhash` on _others_.  External libraries can hash however they please.

Returning to your question, I hope it now makes sense why I would prefer `PersistentVector` to be natively hashable!

spy16 replied:
Ahh Right, that makes sense 😅. Thanks for clarifying.

lthibault replied:
No worries!  I don't think I was very clear the first time around. 😄

spy16 commented on core/vector.go:
```diff
@@ -0,0 +1,20 @@
+// Vector is an ordered collection providing fast random access.
+type Vector interface {
+	Seqable
+	SExpressable
```
I feel, may be we should make the `Seqable` and `SExpressable` optional? For both features we have a sensible fallback.

For `Seqable`: 

```go
func ToSeq(v core.Any) (Seq, error) {
   switch s := v.(type) {
      case Seq:
            return s, nil
      case Seqable:
            return s.Seq()
      case Vector:
            // iterate and create Seq
      default:
            return nil, errors.New("don't know how to create Seq from ??")
    }
}
```

Also, I have a general concern about `SExpressable`. We allow users to change everything about the reader. Including syntax of a `Vector` for example (i.e., I can use `<1 2 3>` to denote a vector in my usage of Slurp). But the value types implementing `SExpressable` use some fixed formatting (e.g., `[1 2 3]` in vector) which now creates a mismatch in what was read as a vector vs what s-expr is emitted.. It's probably not a big issue since usually syntax is fairly similar across LISPs.

lthibault replied:
>I feel, may be we should make the Seqable and SExpressable optional? For both features we have a sensible fallback.

👍 For `SExpressable`. 

I hesitate a _little_  bit to remove `Sequable` from `Vector`,  because a vector is partly defined by the fact that it is a sequence type.  I can't think of a situation where a non-iterable vector is desirable.

Do you have a specific case in mind where forcing `Vector` to be a `Sequable` presents a problem?

> Also, I have a general concern about SExpressable.

This is a good point.  Here are possibilities I currently see:

1. Define some sort of symbol table for use by `repl.Printer`.  This would effectively be the inverse function to the macro table in `Reader`.  Implementation details are a bit unclear to me at this stage.
2. Define some kind of parametrizable factory that knows about delimiters for a given type.
  Think: `NewVectorFactory("[", "]")`.

I'm going out on a limb here, but I think option 2 has the potential to solve a longstanding annoyance of mine in this project:  the fact that the built-in reader macros are tightly coupled to the built-in type implementations.  I would really like to be able to substitute e.g. my own `String` implementation without having to re-implement `readString`.  Here's what that might look like:

```go
package reader

func NewKeywordReader(newString func(string) (core.Any, error)) Macro {
        func (rd *Reader, init rune) (core.Any, error) {
        	beginPos := rd.Position()

	        token, err := rd.Token(-1)
        	if err != nil {
		        return nil, rd.annotateErr(err, beginPos, token)
	        }

        	return newString(s)
        }
}

func NewListReader(newList func(...core.Any) (core.Any, error)) Macro {
        func (rd *Reader, init rune) (core.Any, error) {
		const listEnd = ')'

		beginPos := rd.Position()

		forms := make([]core.Any, 0, 32) // pre-allocate to improve performance on small lists
		if err := rd.Container(listEnd, "list", func(val core.Any) error {
			forms = append(forms, val)
			return nil
		}); err != nil {
			return nil, rd.annotateErr(err, beginPos, "")
		}

		return newList(forms...)
        }
}
```

The macro table could then be assembled in the following way:

```go
	macros: map[rune]Macro{
	'"':  NewStringReader(NewStringFactory("\"", "\"")),
	// ...
	'(':  NewListReader(NewListFactory("(", ")")),
	')':  UnmatchedDelimiter(),
	// ...
},
```

Under the hood `NewStringFactory` would parametrize the `String` type to use double-quotes, and the `List` type to use parens.

Thoughts?  (I haven't thought of this very hard, so I might be overlooking some obvious issues...)

lthibault replied:
>I hesitate a little bit to remove Sequable from Vector, because a vector is partly defined by the fact that it is a sequence type. I can't think of a situation where a non-iterable vector is desirable.

Nevermind, there was some confusion on my end as to the meaning of `Seq`.  I was thinking of `Seq` as an analog to `Iterable`, but it's actually more like `ListSemanticsProvider`.

Mental model updated. ⚡ 🧠 

(Also, 👍 for removing `Sequable` from `Vector` interface.)

spy16 replied:
Seq is still a sequence type. And vector is definitely a sequence. but by embedding seqable we are saying "vector should be convertible to a sequence" and not "vector itself is a sequence" (this is fine since this is our requirement anyway)

Since our requirement is that we should be able to convert a vector to a sequence type, and considering keeping things easy and simple is always a goal, I suggested the ToSeq approach. How it helps?

1. Implementing a custom vector type is easier (since I don't need to implement the seqable part if I don't want to)
2. I can make the vector itself a seq (instead of seqable) if there is a optimal way to look at my vector implementation as a seq. (This might be problematic too since we expect to be able to differentiate vector and a list 😅 so may be wishful thinking)
3. I can make my vector Seqable if I want to optimise it. 

ToSeq will take care of choosing the most appropriate variant and converting it to a Seq. 

Also, we could add logic to convert Go native arrays and slices to a Seq in ToSeq.

spy16 replied:
Regarding the reader/printer, I'm also not sure how to generalize it. But the simplest approach I can think of is to let the user build a printer which relies on type switch to emit lisp source based on their reader setting. 

(Repl package could have one such function based on the built-in types and the reader config repl uses).

lthibault replied:
Re Sequable:  Thanks for the clarifications.  👍 On my end.

Re SExpressable:  should we just shelf this discussion for now and come back to it later?

spy16 replied:
Yep. I think that makes sense. Print is not priority.. also I don't think I will get any time to go through this soon, I went over it once. At a high level looks good to me. feel free to merge when you address these comments...

lthibault replied:
>also I don't think I will get any time to go through this soon [...] feel free to merge when you address these comments.

Will do.   I hope everything is going well on your end. 🙂 

Thanks for the feedback!

spy16 commented on builtin/value_test.go:
```diff
@@ -37,30 +37,31 @@ func TestFloat64(t *testing.T) {
+	assert.Equal(t, want, got)
 }
+
+// func assert(t *testing.T, cond bool, msg string, args ...interface{}) {
```
feel free to remove this completely if not used anymore.

lthibault committed:
Remove commented `assert` function

lthibault committed:
Remove SExpressable from Vector interface.

lthibault committed:
Flesh out vector tests.

lthibault committed:
Split TestSExpr into TestPersistentVector and TestTransientVector subtests.

lthibault committed:
Fix TransientVector.

lthibault committed:
Implement Vector.Pop()
Implement additional transient-based optimizations.
Bugfix in chunkedSeq.

lthibault committed:
Hook Vector up to Reader

lthibault committed:
Bugfix:  support empty vector from VectorBuilder.

lthibault committed:
Add conj example.

lthibault commented:
Ok, ready to go!

I've done the following:

1. Implemented `Pop()`
2. Added extensive unit-testing.  We're north of 90% coverage for all things related to `core.Vector`. 
3. Added transient-based optimizations to `PersistentVector`.  I just went for the low-hanging fruit.  There's plenty of room for improvement.
4. Added `example/conj` just to play around with the vector's `Conj` semantics.  Check it out!
  ```lisp
  Welcome to slurp!
  >> (conj [] 1)
  [1]
  >> (conj (conj [] 1) 2)
  [1 2]
  ```

⏱️ Estimated review time:  15-20 min.
✅ Merge when ready.

lthibault committed:
Fix PersistentVector.Seq docstring.

lthibault committed:
Export builtin.Transientvector.  Remove transient.VectorBuilder.
